Insert custom error VC into WPAuthenticator, overriding inline error message

ctarda opened the pull request:
Closes #3218 
Ref: p91TBi-3Rn-p2

⚠️ Please note this PR is not against develop, but against the feature branch for UL&S 🙇

⚠️⚠️ Please note also that the coloring of the CTAs (Back button, Help button and others) is implemented in #3253 

## Design
<img src="https://user-images.githubusercontent.com/2722505/100716098-338ed400-33f3-11eb-92ee-4bdf959b3c4a.png" width="350"/>

⚠️⚠️⚠️ Please note there is a limitation that, as far as I can tell, we can not work around: at the moment of presenting the error there is no way to know the site address the user entered, so I removed it from the error message.

## Changes

First a gif of the new behaviour:
<img src="https://user-images.githubusercontent.com/2722505/100716301-781a6f80-33f3-11eb-9cc3-107607045ddb.gif" width="350"/>


Screenshots:
| Light mode | Dark mode |
| ---- | ---- |
| <img src="https://user-images.githubusercontent.com/2722505/100716412-a1d39680-33f3-11eb-91c2-08e78fd0adaa.png" width="350"/> | <img src="https://user-images.githubusercontent.com/2722505/100716552-d3e4f880-33f3-11eb-96d6-8003e8b10af6.png" width="350"/> |

The changes to the code are:
* Added a new view model to represent this error
* Pointed WPAuthenticator to the corresponding branch
* Implemented the new methods declared in WPAuthenticatorDelegate, and passed a properly customized error UI to WPAuthenticator.

## How to test
* Check out the branch. Run `bundle exec pod install`
* Log out if necessary
* Attempt to log in with a store address. Use an address that does not correspond to a WP site (i.e. GitHub.com)
* Note: The "Log In With Another Account" button in the error screen should pop to the Login Prologue. The "Enter Another Site" button is deactivated in order to submit the PR, as the functionality to support t is part of a different PR in WPAuthenticator wordpress-mobile/WordPressAuthenticator-iOS#532 . Making progress in two different repos with the desired level of granularity is hard!

Update release notes:
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ctarda committed:
Add view model for error when the site is not a WP site

ctarda committed:
Declare icon for error screen. Add tests

ctarda committed:
Merge branch 'issue/3161-intercept-navigation' into issue/3218-login-error-not-wp

ctarda committed:
Point WPAuthenticator to the branch that allows customising errors

ctarda commented on WooCommerce/Classes/Authentication/AuthenticationManager.swift:
```diff
@@ -183,6 +183,19 @@ extension AuthenticationManager: WordPressAuthenticatorDelegate {
     ///
     func createdWordPressComAccount(username: String, authToken: String) { }
 
+    func shouldDisplayError(_ error: Error) -> Bool {
```
This is the implementation of the two new methods in WPAuthenticatorDelegate. First we signal that we want to handle errors (this should be more granular, and return true only for certain error codes)

ctarda commented on WooCommerce/Classes/Authentication/AuthenticationManager.swift:
```diff
@@ -183,6 +183,19 @@ extension AuthenticationManager: WordPressAuthenticatorDelegate {
+    }
+
+    func handleError(_ error: Error, onCompletion: @escaping (WordPressAuthenticatorResult) -> Void) {
+        let viewModel = NotWPErrorViewModel(siteURL: "http://woocommerce.com")
```
And here we instantiate custom UI and pass it back to WPAuthenticator. WPAuthenticator then will short-circuit its own error handling flow and just present the UI we provide

ctarda replied:
It's probably fine if we just pass a view controller, instead of an instance of WordPressAuthenticatorResult to the completion closure. After all, host apps would only call this method *if* they want to insert custom UI to present the error

ctarda committed:
Merge branch 'issue/3161-intercept-navigation' into issue/3218-login-error-not-wp

ctarda committed:
Merge branch 'feat/unified-login' into issue/3218-login-error-not-wp

# Conflicts:
#	Podfile
#	Podfile.lock

ctarda committed:
Perform first attempt to respond only to the 406 error emitted by WPAuthenticator

ctarda committed:
Use the proper image

ctarda committed:
Use command to pop back one step in the navitation stack

ctarda committed:
Remove custom messages from view model

ctarda committed:
Comment out the call to a command that is not available yet

Garance91540 reviewed (approved):
Looking good Cesar! Is it possible to change "Site" with "Store" on the "Try another site" CTA?

Ecarrion commented on WooCommerce/Classes/Authentication/AuthenticationManager.swift:
```diff
@@ -183,6 +183,25 @@ extension AuthenticationManager: WordPressAuthenticatorDelegate {
+        /// This is the error code that signals that a site is not a WP site.
+        /// It corresponds to WordPressKit.WordPressOrgXMLRPCValidatorError
+        /// However, we do not pull WordPressKit as a dependency
+        guard err.code == 406 else {
```
Do you think it's worth it for us to have those error codes mapped to an internal enum?

That way the could encapsulate that logic to something like:
```swift
func shouldHandleError(_ error: Error) -> Bool {
   return OurNewError(error) == .notWPSite
}
```

ctarda replied:
it makes sense!

Ecarrion commented on WooCommerce/Classes/Authentication/Navigation Exceptions/NotWPErrorViewModel.swift:
```diff
@@ -0,0 +1,65 @@
+    var text: NSAttributedString {
+        let message = NSMutableAttributedString(string: Localization.errorMessage)
+        return message
+    }
```
nit:
```suggestion
    var text: NSAttributedString {
        NSMutableAttributedString(string: Localization.errorMessage)
    }
```

Ecarrion commented on WooCommerce/Classes/Authentication/Navigation Exceptions/NotWPErrorViewModel.swift:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    enum Strings {
+        static let instructionsURLString = "https://docs.woocommerce.com/document/jetpack-setup-instructions-for-the-woocommerce-mobile-app/"
```
I'm not seeing this being used, will it be needed later?

ctarda replied:
Good catch, I'll remove it!

Ecarrion commented on WooCommerce/Classes/Extensions/UIImage+Woo.swift:
```diff
@@ -309,12 +309,18 @@ extension UIImage {
+    /// Login error: no WordPress
+    ///
+    static var loginNoWordPressError: UIImage {
+        return UIImage(named: "woo-wp-no-site")!
```
I think this has to be rotated for RTL layouts right?
```suggestion
        return UIImage(named: "woo-wp-no-site")!.imageFlippedForRightToLeftLayoutDirection()
```

Ecarrion commented on WooCommerce/Classes/Authentication/Navigation Exceptions/NotWPErrorViewModel.swift:
```diff
@@ -0,0 +1,65 @@
+import WordPressUI
+
+
+/// Configuration and actions for an ULErrorViewController, modelling
```
typo: `modelling` -> `modeling`
```suggestion
/// Configuration and actions for an ULErrorViewController, modeling
```

Ecarrion reviewed (commented):
Left some minor comments/suggestions. Also, I'm not sure if this is till up for debate but it seems an unnecessary transition to display an error on another screen only to be forced to go back again. Do you think it will be better to show that error on the current screen?

ctarda commented:
@Garance91540 Do you have any input about @Ecarrion 's suggestion here? : https://github.com/woocommerce/woocommerce-ios/pull/3219#pullrequestreview-542078189

From a purely technical point of view, I don't think WPAuthenticator will let us insert custom UI without a transition to a separate screen. The view where users enter the store address is served by WPAuthenticator. The view where we present the error is served by WooCommerce.

ctarda committed:
Point WPAuthenticator to develop until the beta release propagates to the CDN

ctarda committed:
Update copy of primary action button

ctarda committed:
Uncommet command to pop navigation from primary button

ctarda committed:
Remove a return statement

ctarda committed:
Remove unused string

ctarda committed:
Flip error images for RTL for both no jetpack and no wp error

ctarda committed:
Fix typo

ctarda committed:
Map error codes to a private enumeration

ctarda committed:
Pin WPAuthenticator to 1.31.0-beta.2

ctarda commented:
I think I've covered everything:
* Added a private enumeration to map to error codes emitted by WPAuthenticator
* Various nits and bits (including making the image for the Jetpack error aware or RTL)
* Pinned WPAuthenticator to the beta release that implements support for all of this.
* Updated the copy of the primary action button:

<img src="https://user-images.githubusercontent.com/2722505/100823318-f0823e80-348e-11eb-923a-f44c16d29b79.png" width="350"/>

This is ready for another round!

Ecarrion commented on WooCommerce/Classes/Authentication/AuthenticationManager.swift:
```diff
@@ -344,3 +340,25 @@ extension AuthenticationManager: WordPressAuthenticatorDelegate {
+    /// Maps error codes emitted by WPAuthenticator to a domain error object
+    enum AuthenticationError: Int, Error {
+        case notWPSite = 406
+        case unkown
```
```suggestion
        case unknown
```

Ecarrion reviewed (approved):
Looks good! Just a minor typo.
🚀

ctarda committed:
Fix typo

ctarda commented:
Alrighty, typo in the enumeration fixed! Merging! 🚀
